src/main/java/org/datadog/jenkins/plugins/datadog/clients/HttpClient.java

```diff
@@ -0,0 +1,291 @@
+package org.datadog.jenkins.plugins.datadog.clients;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+import jenkins.model.Jenkins;
+import org.datadog.jenkins.plugins.datadog.DatadogUtilities;
+import org.datadog.jenkins.plugins.datadog.transport.HttpMessage;
+import org.eclipse.jetty.client.HttpProxy;
+import org.eclipse.jetty.client.Origin;
+import org.eclipse.jetty.client.ProxyConfiguration;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.api.Response;
+import org.eclipse.jetty.client.api.Result;
+import org.eclipse.jetty.client.util.BufferingResponseListener;
+import org.eclipse.jetty.client.util.BytesContentProvider;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.util.ssl.SslContextFactory;
+import org.eclipse.jetty.util.thread.QueuedThreadPool;
+
+public class HttpClient {
+
+    private static final org.eclipse.jetty.client.HttpClient CLIENT = buildHttpClient();
+    private static volatile hudson.ProxyConfiguration EFFECTIVE_PROXY_CONFIGURATION;
+
+    private static final Logger logger = Logger.getLogger(HttpClient.class.getName());
+
+    private static org.eclipse.jetty.client.HttpClient buildHttpClient() {
+        BlockingQueue<Runnable> queue = new ArrayBlockingQueue<>(1024);
+        ThreadFactory threadFactory = new ThreadFactory() {
+            final ThreadFactory delegate = Executors.defaultThreadFactory();
+
+            @Override
+            public Thread newThread(final Runnable r) {
+                final Thread result = delegate.newThread(r);
+                result.setName("dd-http-client-" + result.getName());
+                result.setDaemon(true);
+                return result;
+            }
+        };
+
+        QueuedThreadPool threadPool = new QueuedThreadPool(
+                64,
```

Comment:
One of the problems that the previous version of the code had was that it did not make any attempt at managing resources: there was no limit on the number of TCP connections the plugin could open simultaneously.
If there was a network issue that slowed down the responses, and provided the load was high enough, the process could hit the limit of file descriptors it was allowed to open.

One of the goals of this change was to ensure this does not happen. Given that the default `ulimit` value is usually somewhere between 1024 and 4096 in most systems, I considered 64 low enough to not worry about exhausting the FDs.

It can be fine-tuned further, of course, if we want to achieve the highest possible throughput, but in order to calculate the optimal value we'd have to know the number of CPU cores (which we can easily get at runtime in Java) and the ratio between CPU-intensive and IO-intensive work. The latter is difficult: given how little control we have over the environment in which the plugin runs, and how much it varies from customer to customer, I think it's impossible to do proper calculation here.

I'd say 64 is more than any client is ever going to need, provided that the network is healthy. In that regard it mimics previous behaviour (i.e. no limit), and we'd only be able to see the difference if requests stall en masse.

Reply:
Thanks for the explanation!